base_layer/mmr/src/mutable_mmr.rs

```diff
@@ -0,0 +1,182 @@
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{backend::ArrayLike, common::leaf_index, error::MerkleMountainRangeError, Hash, MerkleMountainRange};
+use croaring::Bitmap;
+use digest::Digest;
+
+/// Unlike a pure MMR, which is append-only, in `MutableMmr`, leaf nodes can be marked as deleted.
+///
+/// In `MutableMmr` a roaring bitmap tracks which data have been marked as deleted, and the merklish root is modified
+/// to include the hash of the roaring bitmap.
+///
+/// The `MutableMmr` API maps nearly 1:1 to that of MerkleMountainRange so that you should be able to use it as a
+/// drop-in replacement for the latter in most cases.
+pub struct MutableMmr<D, B>
+where
+    D: Digest,
+    B: ArrayLike<Value = Hash>,
+{
+    pub(crate) mmr: MerkleMountainRange<D, B>,
+    deleted: Bitmap,
+    // The number of leaf nodes in the MutableMmr. Bitmap is limited to 4 billion elements, which is plenty.
+    // [croaring::Treemap] is a 64bit alternative, but this would break things on 32bit systems. A good TODO would be
+    // to select the bitmap backend using a feature flag
+    size: u32,
+}
+
+impl<D, B> MutableMmr<D, B>
+where
+    D: Digest,
+    B: ArrayLike<Value = Hash>,
+{
+    /// Create a new mutable MMR using the backend provided
+    pub fn new(mmr_backend: B) -> MutableMmr<D, B> {
+        let mmr = MerkleMountainRange::new(mmr_backend);
+        MutableMmr {
+            mmr,
+            deleted: Bitmap::create(),
+            size: 0,
+        }
+    }
+
+    /// Return the number of leaf nodes in the `MutableMmr` that have not been marked as deleted.
+    ///
+    /// NB: This is semantically different to `MerkleMountainRange::len()`. The latter returns the total number of
+    /// nodes in the MMR, while this function returns the number of leaf nodes minus the number of nodes marked for
+    /// deletion.
+    #[inline(always)]
+    pub fn len(&self) -> u32 {
```

Comment:
```suggestion
    pub fn len(&self) -> usize {
```
The MMR file uses `usize`, is there a reason that this is different?